File: gateaway/src/limiter/tokenBucket.js
/**
 * Token Bucket Rate Limiter — Lua-backed atomic implementation
 *
 * WHY THIS EXISTS — what sliding window still can't do:
 *
 *   Both fixed and sliding windows ask: "how many requests in the last N seconds?"
 *   They treat every second of the window as equal.
 *
 *   Token bucket asks a different question:
 *   "Does this client have a token to spend right now?"
 *
 * HOW IT WORKS:
 *
 *   Imagine a bucket that holds tokens (max = capacity).
 *   Tokens refill at a fixed rate (e.g. 2 tokens per second).
 *   Each request costs 1 token.
 *   If the bucket is empty → reject the request.
 *
 *   This allows BURSTS — if a client hasn't made requests for a while,
 *   their bucket fills up and they can fire several requests quickly.
 *   But sustained abuse drains the bucket and they get throttled.
 *
 * WHY LUA — the double-spend race condition:
 *
 *   Without atomicity, two gateways both see tokens=1:
 *
 *     Gateway A: HGETALL → tokens=1 → allowed, writes tokens=0
 *     Gateway B: HGETALL → tokens=1 → allowed, writes tokens=0
 *     Both pass. The single token was spent twice.
 *
 *   The Lua script is atomic — Redis processes it as one indivisible command.
 *   Gateway B's script sees tokens=0 (Gateway A already updated it).
 *
 * REDIS DATA STRUCTURE — Hash:
 *
 *   Key:    rl:tb:ip:route:api_data:127.0.0.1
 *   Fields: tokens (current count), lastRefill (timestamp ms)
 */

import { redis } from "../redis/client.js";

/**
 * @param {object} opts
 * @param {string} opts.key              - Redis key for this client+route
 * @param {number} opts.capacity         - Max tokens the bucket can hold
 * @param {number} opts.refillRate       - Tokens added per refillIntervalMs
 * @param {number} opts.refillIntervalMs - How often tokens are added (ms)
 * @param {number} [opts.ttlMs]          - Key TTL for idle cleanup (default: 60s)
 */
export async function consumeTokenBucket({
  key,
  capacity,
  refillRate,
  refillIntervalMs,
  ttlMs = 60 * 1000,
}) {
  const now = Date.now();

  // Single atomic Lua call — replaces the HGETALL → compute → HSET sequence.
  // Returns: [allowed, tokens, msUntilNextToken, capacity]
  const [allowed, tokens, msUntilNextToken] =
    await redis.tokenBucketConsume(
      key,              // KEYS[1]
      now,              // ARGV[1]
      capacity,         // ARGV[2]
      refillRate,       // ARGV[3]
      refillIntervalMs, // ARGV[4]
      ttlMs,            // ARGV[5]
    );

  return {
    allowed:      allowed === 1,
    limit:        capacity,
    current:      tokens,
    remaining:    tokens,
    retryAfterMs: msUntilNextToken > 0 ? msUntilNextToken : refillIntervalMs,
  };
}

File: gateaway/src/limiter/scripts/tokenBucket.lua
--[[
  Atomic Token Bucket Rate Limiter
  =================================
  The race condition this fixes:

    Without Lua — two gateways both read tokens=1 from Redis:
      Gateway A: HGETALL → tokens=1  → allowed, write tokens=0
      Gateway B: HGETALL → tokens=1  → allowed, write tokens=0
      Both allowed. Both wrong. Token was double-spent.

    With this Lua script — Redis serialises all access to the key:
      Gateway A: runs entire script → tokens=1 → allowed, writes tokens=0
      Gateway B: runs entire script → tokens=0 → blocked immediately

  Arguments (ARGV):
    ARGV[1]  now              — current timestamp in ms
    ARGV[2]  capacity         — max tokens the bucket can hold
    ARGV[3]  refillRate       — tokens added per interval
    ARGV[4]  refillIntervalMs — how often tokens are added
    ARGV[5]  ttlMs            — key TTL for idle cleanup

  Keys (KEYS):
    KEYS[1]  — the Redis hash key for this client+route
              Fields: tokens (float), lastRefill (timestamp ms)

  Returns a flat table:
    [1]  allowed          (1 = yes, 0 = no)
    [2]  tokens           (remaining after this request, floored)
    [3]  msUntilNextToken (0 if allowed, ms until 1 token refills if blocked)
    [4]  capacity         (echoed back for header generation)
--]]

local key             = KEYS[1]
local now             = tonumber(ARGV[1])
local capacity        = tonumber(ARGV[2])
local refillRate      = tonumber(ARGV[3])
local refillIntervalMs = tonumber(ARGV[4])
local ttlMs           = tonumber(ARGV[5])

-- Step 1: Load current bucket state
local raw = redis.call("HGETALL", key)

local tokens
local lastRefill

if #raw == 0 then
  -- Brand new client: start with a full bucket
  tokens     = capacity
  lastRefill = now
else
  -- Parse the hash fields (HGETALL returns [field, value, field, value, ...])
  local map = {}
  for i = 1, #raw, 2 do
    map[raw[i]] = raw[i + 1]
  end
  tokens     = tonumber(map["tokens"])     or capacity
  lastRefill = tonumber(map["lastRefill"]) or now
end

-- Step 2: Refill — calculate how many whole intervals have passed
local elapsed          = now - lastRefill
local intervalsElapsed = math.floor(elapsed / refillIntervalMs)
local tokensToAdd      = intervalsElapsed * refillRate

if tokensToAdd > 0 then
  tokens = math.min(capacity, tokens + tokensToAdd)
  -- Advance lastRefill by whole intervals only — preserves partial interval progress
  lastRefill = lastRefill + intervalsElapsed * refillIntervalMs
end

-- Step 3: Try to consume one token
local allowed = 0
if tokens >= 1 then
  tokens  = tokens - 1
  allowed = 1
end

-- Step 4: Persist updated state
redis.call("HSET", key, "tokens", tostring(tokens))
redis.call("HSET", key, "lastRefill", tostring(lastRefill))
redis.call("PEXPIRE", key, ttlMs)

-- How long until next token refills?
local msUntilNextToken = 0
if allowed == 0 then
  msUntilNextToken = refillIntervalMs - (now - lastRefill)
  if msUntilNextToken < 0 then msUntilNextToken = refillIntervalMs end
end

return { allowed, math.floor(tokens), msUntilNextToken, capacity }
